### DISCUSSION TITLE
Connection Lost Handler in Silohost

### DISCUSSION THREAD

**Discussion by JunioPereira** [dotnet/orleans#7177] [General]
hello guys,
 Is possible we have a Cluster Connection Lost Handler in Silohost?
I want know when a server shutdown/ or crashed.

**Comment by ReubenBond** [comment]
You can get notifications from `IClusterMembershipService.MembershipUpdates`, which is `IAsyncEnumerable<ClusterMembershopSnapshot>` and those snapshots are diffable, so you can take any two updates and call `first.CreateUpdate(second)` to get a `ClusterMembershipUpdate` object which contains a list of changes (check the `Status` property).

Something like this:
``` C# 
var previous = _clusterMembershipService.CurrentSnapshot;
await foreach (var snapshot in _clusterMembershipService.MembershipUpdates.WithCancellation(shutdownToken))
{
  var diff = snapshot.CreateUpdate(previous);
  foreach (var change in diff.Changes)
  {
    if (change.Status == SiloStatus.Dead)
    {
      Console.WriteLine($"Server {change.SiloAddress} has been removed");
    }
  }

  previous = snapshot;
}
```

**Reply by JunioPereira** [reply]
nice, thanks. This interface "IClusterMembershipService" is injectable?
